sphinx/registry.py

```diff
@@ -200,24 +206,48 @@ def add_crossref_type(self, directivename, rolename, indextemplate='',
     def add_source_parser(self, suffix, parser):
         # type: (unicode, Type[Parser]) -> None
         logger.debug('[app] adding search source_parser: %r, %r', suffix, parser)
-        if suffix in self.source_parsers:
+        if suffix in self.source_suffix:
             raise ExtensionError(__('source_parser for %r is already registered') % suffix)
+        else:
+            self.source_suffix[suffix] = suffix
+
+        if len(parser.supported) == 0:
+            warnings.warn('Old source_parser has been detected. Please fill Parser.supported '
+                          'attribute: %s' % parser.__name__,
+                          RemovedInSphinx30Warning)
+
+        # create a map from filetype to parser
+        for filetype in parser.supported:
+            if filetype in self.source_parsers:
+                raise ExtensionError(__('source_parser for %r is already registered') %
+                                     filetype)
+            else:
+                self.source_parsers[filetype] = parser
+
+        # also maps suffix to parser
+        #
+        # This allows parsers not having ``supported`` filetypes.
         self.source_parsers[suffix] = parser
 
-    def get_source_parser(self, filename):
-        # type: (unicode) -> Type[Parser]
-        for suffix, parser_class in iteritems(self.source_parsers):
+    def get_filetype(self, filename):
```

Comment:
I'm not sure where is the best to place this method. `sphinx.io` seems also good.